Sources/CoreMetrics/SystemMetrics.swift

```diff
@@ -0,0 +1,235 @@
+            if let rss = metrics.residentMemoryBytes { Gauge(label: self.labels.label(for: \.residentMemoryBytes)).record(rss) }
+            if let start = metrics.startTimeSeconds { Gauge(label: self.labels.label(for: \.startTimeSeconds)).record(start) }
+            if let cpuSeconds = metrics.cpuSeconds { Gauge(label: self.labels.label(for: \.cpuSecondsTotal)).record(cpuSeconds) }
+            if let maxFds = metrics.maxFds { Gauge(label: self.labels.label(for: \.maxFds)).record(maxFds) }
+            if let openFds = metrics.openFds { Gauge(label: self.labels.label(for: \.openFds)).record(openFds) }
+        })
+
+        self.updateSystemMetrics()
+    }
+
+    internal func cancelSystemMetrics() {
+        self.task?.cancel()
+        self.task = nil
+    }
+
+    internal func updateSystemMetrics() {
+        self.queue.asyncAfter(deadline: .now() + self.timeInterval) {
+            guard let task = self.task else { return }
+            task.perform()
+            self.updateSystemMetrics()
+        }
+    }
+}
+
+public struct SystemMetricsLabels {
+    let prefix: String
+    let virtualMemoryBytes: String
+    let residentMemoryBytes: String
+    let startTimeSeconds: String
+    let cpuSecondsTotal: String
+    let maxFds: String
+    let openFds: String
+
+    public init(prefix: String, virtualMemoryBytes: String, residentMemoryBytes: String, startTimeSeconds: String, cpuSecondsTotal: String, maxFds: String, openFds: String) {
+        self.prefix = prefix
+        self.virtualMemoryBytes = virtualMemoryBytes
+        self.residentMemoryBytes = residentMemoryBytes
+        self.startTimeSeconds = startTimeSeconds
+        self.cpuSecondsTotal = cpuSecondsTotal
+        self.maxFds = maxFds
+        self.openFds = openFds
+    }
+
+    func label(for keyPath: KeyPath<SystemMetricsLabels, String>) -> String {
+        return self.prefix + self[keyPath: keyPath]
+    }
+}
+
+public protocol SystemMetrics {
+    init(pid: Int)
+
+    var virtualMemoryBytes: Int? { get }
+    var residentMemoryBytes: Int? { get }
+    var startTimeSeconds: Int? { get }
+    var cpuSeconds: Int? { get }
+    var maxFds: Int? { get }
+    var openFds: Int? { get }
+}
+
+#if os(Linux)
+internal struct LinuxSystemMetrics: SystemMetrics {
+    let virtualMemoryBytes: Int?
+    let residentMemoryBytes: Int?
+    let startTimeSeconds: Int?
+    let cpuSeconds: Int?
+    let maxFds: Int?
+    let openFds: Int?
+
+    private enum SystemMetricsError: Error {
+        case MetricReadError
+    }
+    
+    init(pid: Int) {
+        let pid = "\(pid)"
+        let ticks = _SC_CLK_TCK
+        
+        do {
+            guard
+                let fp = fopen("/proc/\(pid)/stat", "r")
+            else { throw SystemMetricsError.MetricReadError }
+            var buf = [CChar](repeating: CChar(0), count: 1024)
+            while fgets(&buf, 1024, fp) != nil { }
```

Comment:
- this isn't right. This overwrites `buf` in every loop iteration.
- can we please use `buf.withUnsafeBytes { ptr in while fgets(...) }` instead of relying on the dangerous, magic conversion from `&` to `Unsafe*Pointer`?